Storage/src/SigningHelper.php

```diff
@@ -0,0 +1,578 @@
+     */
+    protected function createV2CanonicalRequest(array $canonicalRequest)
+    {
+        return implode("\n", $canonicalRequest);
+    }
+
+    /**
+     * Normalizes and validates an expiration.
+     *
+     * @param Timestamp|\DateTimeInterface|int $expires The expiration
+     * @return int
+     * @throws \InvalidArgumentException If an invalid value is given.
+     */
+    protected function normalizeExpiration($expires)
+    {
+        if ($expires instanceof Timestamp) {
+            $seconds = $expires->get()->format('U');
+        } elseif ($expires instanceof \DateTimeInterface) {
+            $seconds = $expires->format('U');
+        } elseif (is_numeric($expires)) {
+            $seconds = (int) $expires;
+        } else {
+            throw new \InvalidArgumentException('Invalid expiration.');
+        }
+
+        return $seconds;
+    }
+
+    /**
+     * Normalizes and encodes the resource identifier.
+     *
+     * @param string $resource The resource identifier. In form
+     *        `[/]$bucket/$object]`.
+     * @return string The resource, with pieces encoded and prefixed with a
+     *        forward slash.
+     */
+    protected function normalizeResource($resource)
+    {
+        $pieces = explode('/', trim($resource, '/'));
+        array_walk($pieces, function (&$piece) {
+            $piece = rawurlencode($piece);
+        });
+        return '/' . implode('/', $pieces);
+    }
+
+    /**
+     * Fixes the user input options, filters and validates data.
+     *
+     * @param array $options Signed URL configuration options.
+     * @return array
+     * @throws \InvalidArgumentException
+     */
+    protected function normalizeOptions(array $options)
+    {
+        $options += [
+            'method' => 'GET',
+            'cname' => self::DEFAULT_DOWNLOAD_HOST,
+            'contentMd5' => null,
+            'contentType' => null,
+            'headers' => [],
+            'saveAsName' => null,
+            'responseDisposition' => null,
+            'responseType' => null,
+            'keyFile' => null,
+            'keyFilePath' => null,
+            'allowPost' => false,
+            'forceOpenssl' => false,
+            'queryParams' => [],
+            'timestamp' => null
+        ];
+
+        $allowedMethods = ['GET', 'PUT', 'POST', 'DELETE'];
+        $options['method'] = strtoupper($options['method']);
+        if (!in_array($options['method'], $allowedMethods)) {
+            throw new \InvalidArgumentException('$options.method must be one of `GET`, `PUT` or `DELETE`.');
+        }
+
+        if ($options['method'] === 'POST' && !$options['allowPost']) {
+            throw new \InvalidArgumentException(
+                'Invalid method. To create an upload URI, use StorageObject::signedUploadUrl().'
+            );
+        }
+        unset($options['allowPost']);
+
+        // For backwards compatibility, strip protocol from cname.
+        $cnameParts = explode('//', $options['cname']);
```

Comment:
I'm wondering if there may be some edge cases this could miss, what are your thoughts on using a regex instead?

Reply:
Hmm, what edge cases are you thinking that a regex would catch? I considered using a regular expression, but considering that a cname value should be just a bare hostname (i.e. no path), I didn't think it added any extra value.

Reply:
After thinking on it some more, this approach works for me